File: public/api/components.php
<?php

require dirname(__DIR__, 2) . '/vendor/autoload.php';

use App\Database;
use App\ComponentLoader;
use App\ComponentValidator;
use App\ComponentRepository;
use App\AppConfig;

// Auth::requireAuth();

header('Content-Type: application/json');

$method = $_SERVER['REQUEST_METHOD'];

if ($method === 'GET') {
    try {
        $components = ComponentLoader::all();
        $db         = Database::get();
        $repo       = new ComponentRepository($db);
        $output     = [];

        foreach ($components as $component) {
            $activeImage  = $repo->activeImage($component['id']);
            $imageCount   = count($repo->forComponent($component['id']));
            $status       = $repo->generationStatus($component['id'], $component['presets'] ?? []);
            $outputFolder = AppConfig::outputFolder();

            $entry = [
                'id'                 => $component['id'],
                'label'              => $component['label'],
                'description'        => $component['description'] ?? '',
                'presets'            => $component['presets'] ?? [],
                'image_count'        => $imageCount,
                'variants_total'     => $status['total'],
                'variants_generated' => $status['generated'],
                'active_image'       => null,
            ];

            if ($activeImage) {
                $entry['active_image'] = [
                    'assignment_id' => (int) $activeImage['assignment_id'],
                    'id'            => $activeImage['id'],
                    'original_name' => $activeImage['original_name'],
                    'name'          => $activeImage['name'] ?: $activeImage['original_name'],
                    'filename'      => $activeImage['filename'],
                    'width'         => (int) $activeImage['width'],
                    'height'        => (int) $activeImage['height'],
                    'thumb_url'     => 'image-serve.php?file=' . urlencode($activeImage['filename']) . '&w=400&h=250&fit=crop',
                ];
            }

            $output[] = $entry;
        }

        echo json_encode(['components' => $output]);
    } catch (\Exception $e) {
        http_response_code(500);
        echo json_encode(['error' => $e->getMessage()]);
    }
    exit;
}

if ($method === 'POST') {
    $body = json_decode(file_get_contents('php://input'), true);

    if (!isset($body['components']) || !is_array($body['components'])) {
        http_response_code(400);
        echo json_encode(['error' => 'Missing or invalid components array']);
        exit;
    }

    $errors = ComponentValidator::validateComponents($body['components']);

    if (!empty($errors)) {
        http_response_code(422);
        echo json_encode(['error' => 'Validation failed', 'errors' => $errors]);
        exit;
    }

    $path = dirname(__DIR__, 2) . '/config/components.json';
    file_put_contents($path, json_encode(['components' => $body['components']], JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES));

    ComponentLoader::clearCache();

    echo json_encode(['success' => true]);
    exit;
}

http_response_code(405);
echo json_encode(['error' => 'Method not allowed']);

File: internal/handler/components.go
package handler

import (
	"imagemanager/internal/config"
	"imagemanager/internal/model"
	"imagemanager/internal/repo"
	"net/http"
	"net/url"
)

func (a *App) Components(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		components, err := config.LoadComponents(a.AppRoot)
		if err != nil {
			jsonErr(w, 500, err.Error())
			return
		}

		compRepo := repo.NewComponentRepo(a.DB)
		presets, _ := config.LoadPresets(a.AppRoot)
		// Build variant map: presetID → []variantID
		variantMap := make(map[string][]string)
		for _, p := range presets {
			var ids []string
			for _, v := range p.Variants {
				ids = append(ids, v.ID)
			}
			variantMap[p.ID] = ids
		}

		var output []any
		for _, c := range components {
			active, _ := compRepo.ActiveImage(c.ID)
			rows, _ := compRepo.ForComponent(c.ID)
			total, generated, _ := compRepo.GenerationStatusRaw(c.ID, c.Presets, variantMap)

			entry := map[string]any{
				"id":                  c.ID,
				"label":               c.Label,
				"description":         c.Description,
				"presets":             c.Presets,
				"image_count":         len(rows),
				"variants_total":      total,
				"variants_generated":  generated,
				"active_image":        nil,
			}

			if active != nil {
			entry["active_image"] = map[string]any{
				"assignment_id": active.AssignmentID,
				"id":            active.ImageID,
				"original_name": active.OriginalName,
				"name":          active.Name,
				"filename":      active.Filename,
				"width":         active.Width,
				"height":        active.Height,
				"thumb_url":     "image-serve?file=" + url.QueryEscape(active.Filename) + "&w=400&h=250&fit=crop",
			}
		}

			output = append(output, entry)
		}

		if output == nil {
			output = []any{}
		}
		jsonOK(w, map[string]any{"components": output})

	case http.MethodPost:
		var body struct {
			Components []model.Component `json:"components"`
		}
		if err := readJSON(r, &body); err != nil {
			jsonErr(w, 400, "Missing or invalid components array")
			return
		}

		presets, _ := config.LoadPresets(a.AppRoot)
		errs := config.ValidateComponents(body.Components, presets)
		if len(errs) > 0 {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(422)
			jsonOK(w, map[string]any{"error": "Validation failed", "errors": errs})
			return
		}

		if err := config.SaveComponents(a.AppRoot, body.Components); err != nil {
			jsonErr(w, 500, "Failed to save components")
			return
		}
		jsonOK(w, map[string]bool{"success": true})

	default:
		jsonErr(w, 405, "Method not allowed")
	}
}
